From 3d05a287ee3df0f1d0968352da761c1c714b90b1 Mon Sep 17 00:00:00 2001
From: Christian Hewitt <christianshewitt@gmail.com>
Date: Tue, 31 Mar 2026 07:45:05 +0000
Subject: [PATCH 16/23] media: meson: vdec: report MPEG1/2 frame type flags to
 userspace

The MPEG1/2 firmware reports the picture coding type (I/P/B) in the
MREG_PIC_INFO register. Extract this and pass it through to the
V4L2 buffer flags so userspace receives V4L2_BUF_FLAG_KEYFRAME,
V4L2_BUF_FLAG_PFRAME, and V4L2_BUF_FLAG_BFRAME on decoded frames.

Previously the frame type was always passed as 0 (unknown).

Signed-off-by: Christian Hewitt <christianshewitt@gmail.com>
---
 drivers/staging/media/meson/vdec/codec_mpeg12.c | 16 +++++++++++++++-
 1 file changed, 15 insertions(+), 1 deletion(-)

diff --git a/drivers/staging/media/meson/vdec/codec_mpeg12.c b/drivers/staging/media/meson/vdec/codec_mpeg12.c
index 0f45eb403c693..14ff90c7c7eed 100644
--- a/drivers/staging/media/meson/vdec/codec_mpeg12.c
+++ b/drivers/staging/media/meson/vdec/codec_mpeg12.c
@@ -35,6 +35,10 @@
 
 #define PICINFO_PROG		0x00008000
 #define PICINFO_TOP_FIRST	0x00002000
+#define PICINFO_TYPE_MASK	GENMASK(2, 0)
+#define PICINFO_TYPE_I		1
+#define PICINFO_TYPE_P		2
+#define PICINFO_TYPE_B		3
 
 struct codec_mpeg12 {
 	/* Buffer for the MPEG1/2 Workspace */
@@ -158,6 +162,8 @@ static irqreturn_t codec_mpeg12_threaded_isr(struct amvdec_session *sess)
 	u32 is_progressive;
 	u32 buffer_index;
 	u32 field = V4L2_FIELD_NONE;
+	u32 pic_type;
+	u32 type = 0;
 	u32 offset;
 
 	amvdec_write_dos(core, ASSIST_MBOX1_CLR_REG, 1);
@@ -184,10 +190,18 @@ static irqreturn_t codec_mpeg12_threaded_isr(struct amvdec_session *sess)
 			V4L2_FIELD_INTERLACED_TB :
 			V4L2_FIELD_INTERLACED_BT;
 
+	pic_type = pic_info & PICINFO_TYPE_MASK;
+	if (pic_type == PICINFO_TYPE_I)
+		type = 1;
+	else if (pic_type == PICINFO_TYPE_P)
+		type = 2;
+	else if (pic_type == PICINFO_TYPE_B)
+		type = 3;
+
 	codec_mpeg12_update_dar(sess);
 	buffer_index = ((reg & 0xf) - 1) & 7;
 	offset = amvdec_read_dos(core, MREG_FRAME_OFFSET);
-	amvdec_dst_buf_done_idx(sess, buffer_index, offset, field, 0);
+	amvdec_dst_buf_done_idx(sess, buffer_index, offset, field, type);
 
 end:
 	amvdec_write_dos(core, MREG_BUFFEROUT, 0);
